### PR TITLE
[IA-1297] Welder deploy script

### PR DESCRIPTION
**Pull request by rtitle**
JIRA: https://broadworkbench.atlassian.net/browse/IA-1297

This PR:
- Adds `startup.sh` which is deployed as a [GCE startup script](https://cloud.google.com/compute/docs/startupscript) on all clusters
   - Previously this was defined in code but I made it its own file
- Adds a config `deployWelderLabel` which will install welder on all clusters with the given label (we can set it to `saturnAutoCreated` to deploy to Terra clusters)
- Changed startup script installation from cluster _stop_ time to cluster _start_ time, so we can install scripts "just in time"

Unit tests pass, haven't run automation yet. However I've been manually testing in a fiab and it seems to do the right thing. We should do more exhaustive testing of welder functionality.

Have you read [CONTRIBUTING.md](https://github.com/DataBiosphere/leonardo/blob/develop/CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [x] I've documented my API changes in Swagger

In all cases:

- [x] Get a thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: src/main/resources/jupyter/startup.sh [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+set -e -x
+
+##
+# This is a startup script designed to run on Leo-created Dataproc clusters.
+#
+# It starts up Jupyter and Welder processes. It also optionally deploys welder on a
+# cluster if not already installed.
+##
+
+# Templated values
```
Comment (rtitle) [inline src/main/resources/jupyter/startup.sh:12]: I wasn't a big fan of templating but GCE doesn't give a good way to pass args to startup scripts. They suggest reading parameters from instance metadata but that seems unwieldy.

File: src/main/resources/jupyter/startup.sh [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+    # Run welder-docker-compose
+    gcloud auth configure-docker
+    docker-compose -f /etc/welder-docker-compose.yaml up -d
+
+    # Set EVs inside Jupyter container necessary for welder
+    docker exec -it $JUPYTER_SERVER_NAME bash -c "echo $'export WELDER_ENABLED=true\nexport NOTEBOOKS_DIR=$NOTEBOOKS_DIR' >> /home/jupyter-user/.bashrc"
+
+    # Move existing notebooks to new notebooks dir
+    docker exec -it $JUPYTER_SERVER_NAME bash -c "ls -I jupyter.log -I localization.log -I notebooks /home/jupyter-user | xargs -d '\n'  -I file mv file $NOTEBOOKS_DIR"
```
Comment (rtitle) [inline src/main/resources/jupyter/startup.sh]: This `mv`s everything under `~jupyter-user/` to `~jupyter-user/notebooks/`. I didn't want to `cp` because we could potentially run out of disk (depending what data files the user has on their cluster). It's a little scary `mv`ing users' files -- we should definitely test to make sure there are no edge cases. This could also break notebook code that uses absolute paths, but I'm not sure we can avoid that.

File: src/main/resources/reference.conf [hunk 3]
```diff
@@ -28,12 +28,15 @@ dataproc {
   #vpcSubnet = "default"
   customDataprocImage = "projects/broad-dsp-gcr-public/global/images/custom-leo-image-08a4ad8"   # updated 8/26
   dataprocDockerImage = "us.gcr.io/broad-dsp-gcr-public/leonardo-jupyter:5c51ce6935da"           # updated 8/26
+  # Set to deploy welder to clusters with the given label
+  #deployWelderLabel = "saturnVersion"
```
Comment (rtitle) [inline src/main/resources/reference.conf]: Commented for now, so we should be able to deploy this prior to deploying the Terra UI changes, and just enable this config when we're ready.

Comment (Qi77Qi) [reply src/main/resources/reference.conf]: what does this label mean?

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala [hunk 4]
```diff
@@ -1116,4 +1101,39 @@ class LeonardoService(protected val dataprocConfig: DataprocConfig,
...
+    !cluster.welderEnabled && dataprocConfig.deployWelderLabel.exists(cluster.labels.contains)
+  }
+
+  private def enableWelder(cluster: Cluster): IO[Cluster] = {
+    for {
+      _ <- IO(logger.info(s"Deploying welder on cluster ${cluster.projectNameString}"))
+      _ <- Metrics.newRelic.incrementCounterIO("welderDeployed")
+      now <- IO(Instant.now)
+      welderImage = ClusterImage(Welder, dataprocConfig.welderDockerImage, now)
+      _ <- dbRef.inTransactionIO { _.clusterQuery.enableWedler(cluster.id, ClusterImage(Welder, dataprocConfig.welderDockerImage, Instant.now)) }
```
Comment (Qi77Qi) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: should `now` be reused here instead of calling `Instant.now` directly?

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala [hunk 5]
```diff
@@ -1116,4 +1101,39 @@ class LeonardoService(protected val dataprocConfig: DataprocConfig,
...
+      newCluster = cluster.copy(welderEnabled = true, clusterImages = cluster.clusterImages + welderImage)
```
Comment (Qi77Qi) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: will we accidentally add 2 welder images?

File: src/main/resources/jupyter/startup.sh [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+# Start Jupyter
+echo "Starting Jupyter on cluster $GOOGLE_PROJECT / $CLUSTER_NAME..."
+docker exec -d $JUPYTER_SERVER_NAME /bin/bash -c "/etc/jupyter/scripts/run-jupyter.sh $NOTEBOOKS_DIR || /usr/local/bin/jupyter notebook"
+
+# Start welder, if enabled
+if [ "$WELDER_ENABLED" == "true" ] ; then
+    echo "Starting Welder on cluster $GOOGLE_PROJECT / $CLUSTER_NAME..."
+    docker exec -d $WELDER_SERVER_NAME /opt/docker/bin/entrypoint.sh
+fi
```
Comment (Qi77Qi) [inline src/main/resources/jupyter/startup.sh:51]: is it guaranteed that this will only happen one time or it'd be okay to run this script multiple times on a cluster?

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala [hunk 7]
```diff
@@ -1116,4 +1101,39 @@ class LeonardoService(protected val dataprocConfig: DataprocConfig,
...
 
+  private def shouldDeployWelder(cluster: Cluster): Boolean = {
+    !cluster.welderEnabled && dataprocConfig.deployWelderLabel.exists(cluster.labels.contains)
+  }
+
+  private def enableWelder(cluster: Cluster): IO[Cluster] = {
+    for {
+      _ <- IO(logger.info(s"Deploying welder on cluster ${cluster.projectNameString}"))
+      _ <- Metrics.newRelic.incrementCounterIO("welderDeployed")
+      now <- IO(Instant.now)
```
Comment (Qi77Qi) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: I might add `timer: Timer[IO]` as implicit val for this class, and use `timer` here to get `now`

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala [hunk 8]
```diff
@@ -1116,4 +1101,39 @@ class LeonardoService(protected val dataprocConfig: DataprocConfig,
     Set(welderImageOpt, Some(jupyterImage), rstudioImageOpt).flatten
   }
 
+  private def shouldDeployWelder(cluster: Cluster): Boolean = {
+    !cluster.welderEnabled && dataprocConfig.deployWelderLabel.exists(cluster.labels.contains)
+  }
+
+  private def enableWelder(cluster: Cluster): IO[Cluster] = {
+    for {
+      _ <- IO(logger.info(s"Deploying welder on cluster ${cluster.projectNameString}"))
```
Comment (Qi77Qi) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: shall we log cluster name as well

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/DbReference.scala [hunk 9]
```diff
@@ -64,6 +65,10 @@ case class DbReference(private val dbConfig: DatabaseConfig[JdbcProfile])(implic
     import dataAccess.profile.api._
     database.run(f(dataAccess).transactionally.withTransactionIsolation(isolationLevel))
   }
+
+  def inTransactionIO[T](f: (DataAccess) => DBIO[T], isolationLevel: TransactionIsolation = TransactionIsolation.RepeatableRead): IO[T] = {
```
Comment (Qi77Qi) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/DbReference.scala:69]: 👍 for using `IO` here :)

File: src/main/resources/jupyter/startup.sh (hunk 6 above)
Comment (rtitle) [reply src/main/resources/jupyter/startup.sh:51]: This script runs on every cluster startup. It's intended for `DEPLOY_WELDER` to only be enabled one time -- after which `WELDER_ENABLED` should be true. We should test this (maybe I will add a unit test).

File: src/main/resources/reference.conf (hunk 3 above)
Comment (rtitle) [reply src/main/resources/reference.conf]: Terra UI sets this label on all clusters (I think the current version is `6`). I'm not sure when they increase it. I thought this was a good way to distinguish Terra clusters.

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala (hunk 8 above)
Comment (rtitle) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: `projectNameString` logs both project name and cluster name.

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala (hunk 4 above)
Comment (rtitle) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: 👍 good catch

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala (hunk 5 above)
Comment (rtitle) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: It's a `Set` so I don't think so... plus this method is only called for clusters where welder is disabled.

Comment (rtitle) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: Actually changed this to do a DB `upsert` and `filter` on the current set to ensure we don't duplicate it.

File: src/main/resources/reference.conf (hunk 3 above)
Comment (Qi77Qi) [reply src/main/resources/reference.conf]: just want to understand how the label works. How do we get the label on our end?

Comment (rtitle) [reply src/main/resources/reference.conf]: Ah - label is a Leo concept. They're stored in the DB in the `CLUSTER_LABEL` table. Clients can optionally set labels on cluster creation, and use labels to filter `listClusters` results. I happen to know Terra-UI sets this label. :)

**Comment by rtitle** [comment]
Made a couple more changes based on PR feedback; redeploying to a FIAB for further testing.

File: src/main/resources/reference.conf (hunk 3 above)
Comment (Qi77Qi) [reply src/main/resources/reference.conf]: Gotcha! That makes sense

**Review by Qi77Qi — APPROVED** [review]

**Comment by rtitle** [comment]
jenkins retest

**Comment by rtitle** [comment]
jenkins retest
